README.md
# Trak: track chunks of time from the command line

Trak, v0.0.4 (May 1, 2012)
Written by Adam Sharp

## Notice

Trak was recently a Perl script. It has been ported to Ruby, but the
code really looks like it's taken a beating and is definitely NOT what I
want it to ultimately look like. Much more ruby-fying to happen yet, as
well as support for the excellent
[Chronic](https://github.com/mojombo/chronic) gem for natural language
date parsing in the pipeline.

It's now structured as a RubyGem and should hopefully be available on
RubyGems soon.

Stay tuned.

## Description

Trak is a utility that allows you to quickly make a record of how much
time you've spent on various tasks throughout the day.

Work logs are stored in `/Users/yourusername/Documents/Tracker/` with
the format `YEAR-MONTH-DAY-time-log.txt`.

An example work log that trak will create:

    2011-09-01 9:00
    30: nap
    45: procrastinate
    30: uni
    120: trak

## Installation

Trak is available from [RubyGems](https://rubygems.org/gems/trak):

    $ gem install trak

## Usage

    trak [-d|--date DATE] ##<denom> <description>  # => data entry
    trak [-d|--date DATE] [-r|-l]                  # => reporting
    trak [-d|--date DATE] -e                       # => manually edit time log

Where:

* `##` is a decimal signifying how much time has been spent.
* `<denom>` is either hours (`h/hr/hour/hours`) or minutes
  (`m/min/minute/minutes`). `<denom>` is optional and if ommitted,
  Tracker will interpret the time entered as minutes.
* `<description>` is a string containing a brief description of the
  activity.
* `DATE` is a string of the format `YYYY-MM-DD` which represents any
  date. This effects any of Tracker's modes, i.e., insertion, editing or
  reporting.

### Descriptions

You can use either

    $ trak 30 "Foo bar"

or

    $ trak 30 Foo bar

as everything after the first argument is considered the name of the
task.

### Entering time

These are all valid commands:

    $ trak 1h Write trak documentation       # => 1 hour
    $ trak 30min Rewrite trak documentation  # => 30 minutes
    $ trak 4hours Refactor trak              # => 4 hours
    $ trak 15 Lunch                          # => 15 minutes


CHANGELOG.md
# Change log

## v0.0.4 (2012-05-01)

Changed:

* (ca1da2d): Use TRAK_DEBUG environment var instead of --debug
* (d00089a): Moved todos into github issue tracker
* (5cb50db): Fix for report format

Closed:

* (a6669ca): #11 - Refactor command-line specific code into bin/trak +
  refactor Trak module
* (f5fd613): Add -h|--help switch. Fixes #6
* (a113869): Added version info. Fixes #7

## v0.0.3 (2012-04-30)

Changed:

* (3483255): Added exit_err helper method
* (0c0cbf7): Renamed TrackerUtil module to Trak
* (9528123): Restructured lib dir

Fixed:

* (1050905): Fix bug where all time args are truncated to int
* (328b276): Fixed debug wrapper method

## v0.0.2 (2012-04-26)

Changed:

* (8854df7) Refactored currentTimeInMintues to Time#to_minutes
* (d485e0f) Removed currentTimeFormatted and timeToMinutes

## v0.0.1 (2012-04-26)

Fixed (94929f0):

* Exception when tracking first time for day. Removed an erroneous call
  to sprintf instead of strftime. Also ensured use of integers when
  calculating current time in minutes.

## v0.0.0 (2012-04-25)

Tracker is now known as trak, and it's a RubyGem!

Changed:

* Trak has now been ported from Perl to Ruby. The main reason for this
  was for the ability to support the awesome gem Chronic for natural
  language date parsing in an upcoming release. But beyond that,
  RubyGems is fantastic for packaging and distribution and the plan is
  to make this code seriously more modular over the next little while
  (the code is a joke at the moment :O).

## v0.4 (2011-10-17)

Added:

* If the EDITOR environment variable is set, this is now used by default
  to edit a log file with `track -e`.

* The current time is displayed at the end of the report, like so:

        # Today's logged work
        ...
        Hours logged until 9:30 AM (since 9:15 AM). Currently 10:13 AM.

## v0.3

Added:

* Report now calculates the total time spent on each task throughout the
  day
* Time can now be logged as either minutes or hours. The time argument
  can take the format `##<denom>`, where `##` is the amount of time
  spent, and `<denom>` signifies hours (`h/hr/hour/hours`) or minutes
  (`m/min/minute/minutes`).

Changed:

* Report formatting improvements

Fixed:

* Time wrap-around bug where if time logged passed midnight AM/PM would
  display incorrectly.

## v0.2.1

Changed:

* Change log is now found in CHANGELOG.md rather than README.md.

Fixed:

* Fixed a bug where invoking with no arguments would create an entry in
  the log with no time or message. No arguments is a synonym for `-l` or
  `-r`.

## v0.2

Added:

* Tracker now automatically calculates the time you started your day
  based on your first log for that day. This is stored in the time log
  after the date, and used to calculate how far into the day you have
  tracked time.

Changed:

* Everything after the first argument (the amount of time) is now
  considered to be the task name (i.e., the task is no longer truncated
  to the first word).
* Formatting improvements

## v0.1

Added:

* Tracker.pl -l will now tell you what time you've logged time until.
  For example, if your start time is 8:00 AM and you've logged 3 hours,
  it will output the following:

        Hours logged until 11:00 AM (since 8:00 AM).

Changed:

* `-l` and `-r` switches are now synonyms
* Formatting improvements


Tracker.pl
#!/usr/bin/env perl

# command:
#   $ Tracker.pl 30 "I did stuff"
#   $ Tracker.pl 1h "I did twice as much stuff"
#
# format:
#   - First arg is how much time spent
#   - Second arg is a description

use Getopt::Long;
Getopt::Long::Configure ("bundling");

# place where data is stored
$datadir = "$ENV{'HOME'}/Documents/Tracker/";
`mkdir -p $datadir`;

# define command line options
my $do_report = 0;
my $do_edit = 0;
my $date_arg = '';
my $RETVAL = GetOptions('l|r|report' => \$do_report,
                        'd|date=s' => \$date_arg,
                        'e|edit' => \$do_edit);

# if there were invalid options, exit
if (!$RETVAL) {
    exit($RETVAL);
}

# all valid options have been processed, so figure out which mode
# we're in...
my $MODE = '';

# if we found a -r or -l option, ignore everything else
if ($do_report) {
    $MODE = 'report';
}
# now check if the user wants edit mode
elsif ($do_edit) {
    $MODE = 'edit';
}
# if there are still unprocessed args (that didn't look like switches),
# we're in insert mode
elsif (@ARGV > 0) {
    $MODE = 'insert';
}
# if all else fails, there were probably no args to begin with, so we're
# in report mode
else {
    $MODE = 'report';
}

($day, $month, $year) = (localtime)[3,4,5];
$today = sprintf("%04d-%02d-%02d", $year + 1900, $month + 1, $day);
$fdate = '';

# did the user supply a date argument that isn't today?
if ($date_arg ne $today) {
    $fdate = $date_arg;
}
# otherwise use today's date, formatted, and set date_arg to be false
else {
    $fdate = $today;
    $date = '';
}

# set the output file name
$filename = $datadir.$fdate."-time-log.txt";

if ($MODE eq 'report') {
    if (-e $filename) {
        # open the file and get it as an array
        open REPORTFILE, "< $filename" or die "Couldn't open $filename: $!";
        @file = <REPORTFILE>;
        
        # The keys for each hash are the titles of the various tasks logged.
        # The values are the total time spent on the task.
        my %personal, %work;
        
        # find the start time for the day we're reporting on
        @firstLine = split(/\ /, $file[0]);
        my $startTime = $firstLine[1];
        
        # process each line of the file
        foreach $line (@file[1..@file-1]) {
            chomp($line);
            $line =~ /^(\d+)\:\s+(.+)$/;
            $minutes = $1;
            $text = $2;
            if (!($text =~ /personal|uni|lunch|home/)) {
                $work{$text} += $minutes;
            }
            else {
                $personal{$text} += $minutes;
            }
        }
        
        # print the report
        if ($date_arg) {
            print "# Logged work for $fdate\n";
        }
        else {
            print "# Today's logged work\n";
        }
        my $workTotal = printSubReport(\%work, "Work");
        my $personalTotal = printSubReport(\%personal, "Personal");
        
        $newTimeString = to12HourTime(newTimeWithMinutes($startTime, $workTotal + $personalTotal));
        print "Hours logged until ".$newTimeString." (since ".to12HourTime($startTime)."). ";
        
        # if we're reporting for today, print the current time
        if (!$date_arg) {
            print "Currently " . to12HourTime(currentTimeFormatted()) . ".";
        }
        print "\n";
    }
    else {
        if ($date_arg) {
            print STDERR "No time log for $fdate. Track some time first.\n";
        }
        else {
            print STDERR "No time log for today. Track some time first.\n";
        }
    }
}
elsif ($MODE eq 'edit') {
    if (-e $filename) {
        if (exists $ENV{'EDITOR'}) {
            exec("$ENV{'EDITOR'} $filename");
        }
        else {
            `open $filename`;
        }
        exit(0);
    }
    else {
        die "Couldn't open $filename: $!";
    }
}
elsif ($MODE eq 'insert') {
    if ($date_arg) {
        print "WARNING: Adding time to a day other than today is not recommended.\n";
        print "Continue? (y/n) ";
        $input = <STDIN>;
        if (!($input =~ /^y(es)?/i)) {
            print STDERR "Timelog update cancelled.\n";
            exit(1);
        }
    }
    
    # process arguments
    my $minutes = processTimeArgument($ARGV[0]);
    my $message = join(" ", @ARGV[1..@ARGV-1]);
    
    # open the output file
    if (!-e $filename) {
        open TIMELOG, ">> $filename" or die "Couldn't open $filename: $!";
        my $currentTimeInMinutes = timeToMinutes(currentTimeFormatted());
        my $startTime = minutesToTime(nearest15Minutes($currentTimeInMinutes - $minutes));
        print TIMELOG "$fdate $startTime\n";
    }
    else {
        open TIMELOG, ">> $filename" or die "Couldn't open $filename: $!";
    }
    
    # print the logged time
    print TIMELOG "$minutes: $message\n";
}
else {
    die("Couldn't determine the correct mode (I was given '$MODE'):$!");
}







# expects a single argument - the time argument in the format ##m or ##h
# if argument has no m/h qualifier, assume m
# returns a number of minutes
sub processTimeArgument
{
    if (@_ == 1) {
        my $minutes;
        if ($_[0] =~ /^(\d*\.?\d+)((m|min|minute|minutes)|(h|hr|hour|hours))?$/i) {
            my $time = $1, $modifier = $2;
            if ($modifier =~ /h.*/) {
                $minutes = $time * 60;
            }
            else {
                $minutes = $time;
            }
            
            # check enough time has been logged
            if ($minutes < 15) {
                print STDERR "You must log at least 15 minutes.\n";
                exit(1);
            }
            
            return nearest15Minutes(nearestInt($minutes));
        }
        else {
            print STDERR "Incorrectly formatted argument.\n";
            exit(1);
        }
    }
}

# expects a number of minutes
# if less than 60 returns the number with an "m"
# otherwise converts to hours and adds an "h"
sub timeString
{
    return $_[0] >= 60 ? $_[0]/60 . "h" : $_[0] . "m";
}

# expects a hash of tasks mapped to time spent, and a sub-report name
#   (e.g., work, personal)
# prints a formatted sub-report
# returns the total hours worked
sub printSubReport
{
    my %report = %{$_[0]};
    my $name = $_[1], my $total;
    if (%report > 0) {
        foreach (values(%report)) { $total += $_; }
        print "# $name time (".timeString($total).")\n";
        while (($task, $timeSpent) = each(%report)) {
            print "=> " . timeString($timeSpent) . ": $task\n";
        }
    }
    return $total;
}

sub newTimeWithMinutes
{
    my $len = @_;
    if ($len == 2) {
        my @startTime = split(":", $_[0]);
        my $startMinutes = $startTime[0] * 60 + $startTime[1];
        my $endMinutes = $startMinutes + $_[1];
        my $newHours = int($endMinutes / 60);
        my $newMinutes = $endMinutes % 60;
        my $newFormatted = sprintf("%d:%02d", $newHours, $newMinutes);
        return $newFormatted;
    }
}

# expects a time string formatted HH24:MM
sub timeToMinutes
{
    my $len = @_;
    if ($len == 1) {
        my @theTime = split(":", $_[0]);
        my $minutes = $theTime[0] * 60 + $theTime[1];
        return $minutes;
    }
}

# expects an integer
sub minutesToTime
{
    my $len = @_;
    if ($len == 1) {
        my $totalMinutes = $_[0];
        my $minutes = sprintf("%02d", $totalMinutes % 60);
        my $hours = int($totalMinutes / 60);
        return "$hours:$minutes";
    }
}

sub currentTimeFormatted
{
    (my $currentMinutes, my $currentHours) = (localtime)[1,2];
    return $currentHours . ":" . sprintf("%02d", $currentMinutes);
}

# expects an integer which is the amount of minutes logged
sub startTimeInMinutes
{
    my $currentTimeInMinutes = timeToMinutes(currentTimeFormatted());
    my $rounded = nearest15Minutes($currentTimeInMinutes);
    return $rounded - $_[0];
}

sub nearestInt
{
    my $integerPart = int($_[0]);
    my $decimalPart = $_[0] - $integerPart;
    return $decimalPart >= 0.5 ? $integerPart + 1 : $integerPart;
}

# expects an integer
sub nearest15Minutes
{
    my $len = @_;
    if ($len == 1) {
        my $remainder = $_[0] % 15;
        my $basetime = $_[0] - $remainder;
        if ($remainder > 7) {
            return $basetime + 15;
        }
        else {
            return $basetime;
        }
    }
}

sub to12HourTime
{
    my @time = split(":", $_[0]);
    my $suffix;
    
    # if we've passed midnight, wrap around
    while ($time[0] >= 24) { $time[0] -= 24; }
    
    if ($time[0] >= 12) {
        if ($time[0] > 12) { $time[0] -= 12; }
        $suffix = "PM";
    }
    else {
        if ($time[0] == 0) {
            $time[0] = 12;
        }
        $suffix = "AM";
    }
    chomp($time[1]);
    sprintf("%02d", $time[1]);
    return join(":", @time)." $suffix";
}
